README
NAME
    Pod::Spell - a formatter for spellchecking Pod

VERSION
    version 1.15

SYNOPSIS
            use Pod::Spell;
            Pod::Spell->new->parse_from_file( 'File.pm' );

            Pod::Spell->new->parse_from_filehandle( $infile, $outfile );

    Also look at podspell

            % perl -MPod::Spell -e "Pod::Spell->new->parse_from_file(shift)" Thing.pm |spell |fmt

    ...or instead of piping to spell or "ispell", use ">temp.txt", and open
    temp.txt in your word processor for spell-checking.

DESCRIPTION
    Pod::Spell is a Pod formatter whose output is good for spellchecking.
    Pod::Spell rather like Pod::Text, except that it doesn't put much effort
    into actual formatting, and it suppresses things that look like Perl
    symbols or Perl jargon (so that your spellchecking program won't
    complain about mystery words like "$thing" or ""Foo::Bar"" or
    "hashref").

    This class provides no new public methods. All methods of interest are
    inherited from Pod::Parser (which see). The especially interesting ones
    are "parse_from_filehandle" (which without arguments takes from STDIN
    and sends to STDOUT) and "parse_from_file". But you can probably just
    make do with the examples in the synopsis though.

    This class works by filtering out words that look like Perl or any form
    of computerese (like "$thing" or ""N>7"" or ""@{$foo}{'bar','baz'}"",
    anything in C<...> or F<...> codes, anything in verbatim paragraphs
    (code blocks), and anything in the stopword list. The default stopword
    list for a document starts out from the stopword list defined by
    Pod::Wordlist, and can be supplemented (on a per-document basis) by
    having "=for stopwords" / "=for :stopwords" region(s) in a document.

METHODS
  new
  command
  interior_sequence
  textblock
  verbatim
  stopwords
            $self->stopwords->isa('Pod::WordList'); # true

ENCODINGS
    Pod::Parser, which Pod::Spell extends, is extremely naive about
    character encodings. The "parse_from_file" method does not apply any
    PerlIO encoding layer. If your Pod file is encoded in UTF-8, your data
    will be read incorrectly.

    You should instead use "parse_from_filehandle" and manage the input and
    output layers yourself.

            binmode($_, ":utf8") for ($infile, $outfile);
            $my ps = Pod::Spell->new;
            $ps->parse_from_filehandle( $infile, $outfile );

    If your output destination cannot handle UTF-8, you should set your
    output handle to Latin-1 and tell Pod::Spell to strip out words with
    wide characters.

            binmode($infile, ":utf8");
            binmode($outfile, ":encoding(latin1)");
            $my ps = Pod::Spell->new( no_wide_chars => 1 );
            $ps->parse_from_filehandle( $infile, $outfile );

ADDING STOPWORDS
    You can add stopwords on a per-document basis with "=for stopwords" /
    "=for :stopwords" regions, like so:

      =for stopwords  plok Pringe zorch   snik !qux
      foo bar baz quux quuux

    This adds every word in that paragraph after "stopwords" to the stopword
    list, effective for the rest of the document. In such a list, words are
    whitespace-separated. (The amount of whitespace doesn't matter, as long
    as there's no blank lines in the middle of the paragraph.) Plural forms
    are added automatically using Lingua::EN::Inflect. Words beginning with
    "!" are *deleted* from the stopword list -- so "!qux" deletes "qux" from
    the stopword list, if it was in there in the first place. Note that if a
    stopword is all-lowercase, then it means that it's okay in *any* case;
    but if the word has any capital letters, then it means that it's okay
    *only* with *that* case. So a Wordlist entry of "perl" would permit
    "perl", "Perl", and (less interestingly) "PERL", "pERL", "PerL", et
    cetera. However, a Wordlist entry of "Perl" catches only "Perl", not
    "perl". So if you wanted to make sure you said only "Perl", never
    "perl", you could add this to the top of your document:

      =for stopwords !perl Perl

    Then all instances of the word "Perl" would be weeded out of the
    Pod::Spell-formatted version of your document, but any instances of the
    word "perl" would be left in (unless they were in a C<...> or F<...>
    style).

    You can have several "=for stopwords" regions in your document. You can
    even express them like so:

      =begin stopwords

      plok Pringe zorch

      snik !qux

      foo bar
      baz quux quuux

      =end stopwords

    If you want to use E<...> sequences in a "stopwords" region, you have to
    use ":stopwords", as here:

      =for :stopwords
      virtE<ugrave>

    ...meaning that you're adding a stopword of "virtù". If you left the ":"
    out, that would mean you were adding a stopword of "virtE<ugrave>" (with
    a literal E, a literal <, etc), which will have no effect, since any
    occurrences of virtE<ugrave> don't look like a normal human-language
    word anyway, and so would be screened out before the stopword list is
    consulted anyway.

BUGS AND LIMITATIONS
  finding stopwords defined with "=for"
    Pod::Spell makes a single pass over the POD. Stopwords must be added
    before they show up in the POD.

  finding the wordlist
    Pod::Spell uses File::ShareDir::ProjectDistDir if you're getting errors
    about the wordlist being missing, chances are it's a problem with its
    heuristics. Set "PATH_ISDEV_DEBUG=1" or "PATH_FINDDEV_DEBUG=1", or both
    in your environment for debugging, and then file a bug with
    File::ShareDir::ProjectDistDir if necessary.

HINT
    If you feed output of Pod::Spell into your word processor and run a
    spell-check, make sure you're *not* also running a grammar-check --
    because Pod::Spell drops words that it thinks are Perl symbols, jargon,
    or stopwords, this means you'll have ungrammatical sentences, what with
    words being missing and all. And you don't need a grammar checker to
    tell you that.

SEE ALSO
    Pod::Wordlist

    Pod::Parser

    podchecker also known as Pod::Checker

    perlpod, perlpodspec

BUGS
    Please report any bugs or feature requests on the bugtracker website
    https://github.com/xenoterracide/pod-spell/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

CONTRIBUTORS
    *   David Golden <dagolden@cpan.org>

    *   Kent Fredric <kentfredric@gmail.com>

    *   Olivier Mengué <dolmen@cpan.org>

AUTHORS
    *   Sean M. Burke <sburke@cpan.org>

    *   Caleb Cushing <xenoterracide@gmail.com>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2014 by Caleb Cushing.

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)



Changes
Revision history Pod-Spell

1.15      2014-02-28
	- Convert to strict mode of File::ShareDir::ProjectDistDir (KENTNL)
	- change to use Path::Tiny->lines_utf8

1.14      2014-02-12
	- Discontinue use of File::Slurp, use Path::Tiny->lines instead
	  GH #15 ( Ether )

1.13      2013-11-02
	- fix encoding issues issues in pod for perldoc

1.12      2013-10-17
	- improve documentation

1.11      2013-10-17
	- fix shebang to /usr/bin/perl for EUMM replacment GH #13 (dolmen)

1.10      2013-09-27
	[FEATURES]
	- added 'no_wide_chars' option to strip words with such characters
	  from the output.  This may help spellcheckers that can't cope
	  with UTF-8
	[CHANGED]
	- improved punctuation stripping algorithm for better word
	  isolation (example C<< "hello". >> now is found as "hello")
	- strips all word that do not have at least one \w character
	  as a spellchecker isn't likely to do anything useful with them
	[DOCUMENTED]
	- added recommendations for dealing with character encoding

1.09      2013-09-25
	[FIXED]
	- stopwords with 's are learned without 's to match how they are
	  checked
	[CHANGED]
	- leading and trailing punctuation is entirely stripped since the
	  spellchecker won't care anyway.  Trailing periods are *not*
	  stripped as they might be abbreviations ("Ph.D."), but such words
	  are checked against the stoplist both with and without
	  trailing periods to account for abbreviations and words at the
	  end of a sentence.

1.08      2013-09-24
	[FIXED]
	- hyphenated word parts are also checked against the stoplist
	  and stripped (dagolden) GH #9
	[DOCUMENTED]
	- noted that stopwords must be added before words appear in Pod

1.07      2013-09-14
	- add many words
	[FEATURES]
	- Pod::Wordlist is an Object ( dagolden )
	- Lingua::EN::Inflect is used to reduce requiring plurals in the list
	  ( dagolden )
	- the wordlist is now a sharedir file
	[DEPRECATED]
	- accessing %Pod::Wordlist::Wordlist directly

1.06      2013-09-08
	- import stopwords from Pod::Wordlist::hanekomu except names

1.05      2013-07-07
	- add stopwords

1.04      2013-05-08
	- fix tests to work on older perls

1.03      2013-05-08
	- update copyright info
	- require 5.8 (let me know if this is a problem for you)
	- improve tests
	- hopefully fix failing test on windows
	- improve code readability
	- remove DEBUG for private _is_debug method

1.02      2013-05-07
	- convert to dzil and adopt

1.01	2001-10-27
	- first release version ( SBURKE )


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.013.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;

use File::ShareDir::Install;
install_share dist => "share\/dist\/Pod\-Spell";


my %WriteMakefileArgs = (
  "ABSTRACT" => "a formatter for spellchecking Pod",
  "AUTHOR" => "Sean M. Burke <sburke\@cpan.org>, Caleb Cushing <xenoterracide\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "File::ShareDir::Install" => "0.03"
  },
  "DISTNAME" => "Pod-Spell",
  "EXE_FILES" => [
    "bin/podspell"
  ],
  "LICENSE" => "artistic_2",
  "NAME" => "Pod::Spell",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::Tiny" => 0,
    "File::ShareDir::ProjectDistDir" => "1.000",
    "Lingua::EN::Inflect" => 0,
    "Pod::Escapes" => 0,
    "Pod::Parser" => 0,
    "Text::Wrap" => 0,
    "base" => 0,
    "constant" => 0,
    "locale" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "List::Util" => 0,
    "Test::Deep" => 0,
    "Test::More" => 0,
    "utf8" => 0,
    "version" => 0
  },
  "VERSION" => "1.15",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Class::Tiny" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::ShareDir::ProjectDistDir" => "1.000",
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "File::Temp" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Lingua::EN::Inflect" => 0,
  "List::Util" => 0,
  "Pod::Escapes" => 0,
  "Pod::Parser" => 0,
  "Test::Deep" => 0,
  "Test::More" => 0,
  "Text::Wrap" => 0,
  "base" => 0,
  "constant" => 0,
  "locale" => 0,
  "strict" => 0,
  "utf8" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}




lib/Pod/Wordlist.pm
package Pod::Wordlist;
use strict;
use warnings;
use Lingua::EN::Inflect 'PL';
use File::ShareDir::ProjectDistDir 1.000
	dist_file => defaults => { pathtiny => 1 , strict => 1 };

use Class::Tiny {
    wordlist  => \&_copy_wordlist,
    _is_debug => 0,
    no_wide_chars => 0,
};

use constant MAXWORDLENGTH => 50; ## no critic ( ProhibitConstantPragma )

our $VERSION = '1.15'; # VERSION

our %Wordlist; ## no critic ( Variables::ProhibitPackageVars )

sub _copy_wordlist { return { %Wordlist } }

foreach ( dist_file('Pod-Spell', 'wordlist')->lines_utf8({ chomp => 1 })) {
	$Wordlist{$_} = 1;
	$Wordlist{PL($_)} = 1;
}


sub learn_stopwords {
	my ( $self, $text ) = @_;
	my $stopwords = $self->wordlist;

	while ( $text =~ m<(\S+)>g ) {
		my $word = $1;
		if ( $word =~ m/^!(.+)/s ) {
			# "!word" deletes from the stopword list
			my $negation = $1;
			# different $1 from above
			delete $stopwords->{$negation};
			delete $stopwords->{PL($negation)};
			print "Unlearning stopword <$negation>\n" if $self->_is_debug;
		}
		else {
			$word =~ s{'s$}{}; # we strip 's when checking so strip here, too
			$stopwords->{$word} = 1;
			$stopwords->{PL($word)} = 1;
			print "Learning stopword   <$word>\n" if $self->_is_debug;
		}
	}
	return;
}


sub is_stopword {
	my ($self, $word) = @_;
	my $stopwords = $self->wordlist;
	if ( exists $stopwords->{$word} or exists $stopwords->{ lc $word } ) {
		print "  Rejecting   <$word>\n" if $self->_is_debug;
		return 1;
	}
	return;
}


sub strip_stopwords {
	my ($self, $text) = @_;

	# Count the things in $text
	print "Content: <", $text, ">\n" if $self->_is_debug;

	my @words = grep { length($_) < MAXWORDLENGTH } split " ", $text;

	for ( @words ) {
		print "Parsing word: <$_>\n" if $self->_is_debug;
		# some spellcheckers can't cope with anything but Latin1
		$_ = '' if $self->no_wide_chars && /[^\x00-\xFF]/;

		# strip leading punctuation
		s/^[\(\[\{\'\"\:\;\,\?\!\.]+//;

		# keep everything up to trailing punctuation, not counting
		# periods (for abbreviations like "Ph.D."), single-quotes
		# (for contractions like "don't") or colons (for package
		# names like "Foo::Bar")
		s/^([^\)\]\}\"\;\,\?\!]+).*$/$1/;

		# strip trailing single-quote, periods or colons; after this
		# we have a word that could have internal periods or quotes
		s/[\.\'\:]+$//;

		# strip possessive
		s/'s$//i;

		# zero out variable names or things with internal symbols,
		# since those are probably code expressions outside a C<>
		my $is_sigil   = /^[\&\%\$\@\:\<\*\\\_]/;
		my $is_strange = /[\%\^\&\#\$\@\_\<\>\(\)\[\]\{\}\\\*\:\+\/\=\|\`\~]/;
		$_ = '' if $is_sigil || $is_strange;

		# stop if there are no "word" characters left; if it's just
		# punctuation that we didn't happen to strip or it's weird glyphs,
		# the spellchecker won't do any good anyway
		next unless /\w/;

		print "  Checking as <$_>\n" if $self->_is_debug;

		# replace it with any stopword or stopword parts stripped
		$_ = $self->_strip_a_word($_);

		print "  Keeping as  <$_>\n" if $_ && $self->_is_debug;
	}

	return join(" ", grep { length } @words );
}

sub _strip_a_word {
	my ($self, $word) = @_;
	my $remainder;

	# internal period could be abbreviations, so check with
	# trailing period restored and drop or keep on that basis
	if ( /\./ ) {
		my $abbr = "$word.";
		$remainder = $self->is_stopword($abbr) ? '' : $abbr;
	}
	# try word as-is, including possible hyphenation vs stoplist
	elsif ($self->is_stopword($word) ) {
		$remainder = '';
	}
	# check individual parts of hyphenated word, keep whatever isn't a
	# stopword as individual words
	elsif ( $word =~ /-/ ) {
		my @keep;
		for my $part ( split /-/, $word ) {
			push @keep, $part if ! $self->is_stopword( $part );
		}
		$remainder = join(" ", @keep) if @keep;
	}
	# otherwise, we just keep it
	else {
		$remainder = $word;
	}
	return $remainder;
}

1;

# ABSTRACT: English words that come up in Perl documentation

__END__

=pod

=encoding UTF-8

=head1 NAME

Pod::Wordlist - English words that come up in Perl documentation

=head1 VERSION

version 1.15

=head1 DESCRIPTION

Pod::Wordlist is used by L<Pod::Spell|Pod::Spell>, providing a set of words
that are English jargon words that come up in Perl documentation, but which are
not to be found in general English lexicons.  (For example: autovivify,
backreference, chroot, stringify, wantarray.)

You can also use this wordlist with your word processor by just
pasting C<share/wordlist>'s content into your wordprocessor, deleting
the leading Perl code so that only the wordlist remains, and then
spellchecking this resulting list and adding every word in it to your
private lexicon.

=head1 ATTRIBUTES

=head2 wordlist

	ref $self->wordlist eq 'HASH'; # true

This is the instance of the wordlist

=head2 no_wide_chars

If true, words with characters outside the Latin-1 range C<0x00> to C<0xFF> will
be stripped like stopwords.

=head1 METHODS

=head2 learn_stopwords

    $wordlist->learn_stopwords( $text );

Modifies the stopword list based on a text block. See the rules
for <adding stopwords|Pod::Spell/ADDING STOPWORDS> for details.

=head2 is_stopword

	if ( $wordlist->is_stopword( $word ) ) { ... }

Returns true if the word is found in the stopword list.

=head2 strip_stopwords

    my $out = $wordlist->strip_stopwords( $text );

Returns a string with space separated words from the original
text with stopwords removed.

=head1 WORDLIST

Note that the scope of this file is only English, specifically American
English.  (But you may find in useful to incorporate into your own
lexicons, even if they are for other dialects/languages.)

remove any q{'s} before adding to the list.

The list should be sorted and uniqued. The following will work (with GNU
Coreutils ).

	sort share/wordlist -u > /tmp/sorted && mv /tmp/sorted share/wordlist

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/xenoterracide/pod-spell/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHORS

=over 4

=item *

Sean M. Burke <sburke@cpan.org>

=item *

Caleb Cushing <xenoterracide@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Caleb Cushing.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


lib/Pod/Spell.pm
package Pod::Spell;
use 5.008;
use strict;
use warnings;

our $VERSION = '1.15'; # VERSION

use base 'Pod::Parser';

use Pod::Wordlist;
use Pod::Escapes ('e2char');
use Text::Wrap   ('wrap');

# We don't need a very new version of Text::Wrap, altho they are nicer.
$Text::Wrap::huge = 'overflow';  ## no critic ( Variables::ProhibitPackageVars )

use locale;                      # so our uc/lc works right
use Carp;

#==========================================================================
#
#  Override some methods
#

sub new {
	my ( $class, %args ) = @_;

	my $new = $class->SUPER::new( %args );

	$new->{'region'} = [];

	$new->{no_wide_chars} = $args{no_wide_chars};

	$new->{debug} = $args{debug} || $ENV{PERL_POD_SPELL_DEBUG};

	$new->{stopwords} = $args{stopwords} || Pod::Wordlist->new(
		_is_debug => $new->{debug}, no_wide_chars => $args{no_wide_chars}
	);

	return $new;
}

sub stopwords { my $self = shift; return $self->{stopwords} }

sub verbatim { return ''; }    # totally ignore verbatim sections

#----------------------------------------------------------------------

sub _is_debug {
	my $self = shift;

	return $self->{debug} ? 1 : 0;
}

#----------------------------------------------------------------------

sub textblock {
	my ( $self, $paragraph ) = @_;

	if ( @{ $self->{'region'} } ) {

		my $last_region ## no critic ( ProhibitAmbiguousNames )
			= $self->{'region'}[-1];

		if ( $last_region eq 'stopwords' ) {
			$self->stopwords->learn_stopwords($paragraph);
			return;
		}
		elsif ( $last_region eq ':stopwords' ) {
			$self->stopwords->learn_stopwords( $self->interpolate($paragraph) );

			# I guess that'd work.
			return;
		}
		elsif ( $last_region !~ m/^:/s ) {
			printf "Ignoring a textblock because inside a %s region.\n",
				$self->{'region'}[-1] if $self->_is_debug;
			return;
		}

		# else fall thru, as with a :footnote region or something...
	}
	$self->_treat_words( $self->interpolate($paragraph) );
	return;
}

sub command { ## no critic ( ArgUnpacking)
	# why do I have to shift these?
	my ( $self, $command, $text ) = ( shift, shift, @_ );

	return if $command eq 'pod';

	if ( $command eq 'begin' )
	{            ## no critic ( ControlStructures::ProhibitCascadingIfElse )
		my $region_name;

		#print "BEGIN <$_[0]>\n";
		if ( $text =~ m/^\s*(\S+)/s ) {
			$region_name = $1;
		}
		else {
			$region_name = 'WHATNAME';
		}
		print "~~~~ Beginning region \"$region_name\" ~~~~\n"
			if $self->_is_debug;
		push @{ $self->{'region'} }, $region_name;

	}
	elsif ( $command eq 'end' ) {
		pop @{ $self->{'region'} };    # doesn't bother to check

	}
	elsif ( $command eq 'for' ) {
		if ( $text =~ s/^\s*(\:?)stopwords\s*(.*)//s ) {
			my $para = $2;
			$para = $self->interpolate($para) if $1;
			print "Stopword para: <$2>\n" if $self->_is_debug;
			$self->stopwords->learn_stopwords($para);
		}
	}
	elsif ( @{ $self->{'region'} } ) {    # TODO: accept POD formatting
		                                  # ignore
	}
	elsif ($command eq 'head1'
		or $command eq 'head2'
		or $command eq 'head2'
		or $command eq 'head3'
		or $command eq 'item' )
	{
		my $out_fh = $self->output_handle();
		print $out_fh "\n";
		$self->_treat_words( $self->interpolate(shift) );

		#print $out_fh "\n";
	}
	return;
}

#--------------------------------------------------------------------------

sub interior_sequence { ## no critic ( Subroutines::RequireFinalReturn )
	my ( $self, $command, $seq_arg ) = @_;

	return '' if $command eq 'X' or $command eq 'Z';

	# Expand escapes into the actual character now, carping if invalid.
	if ( $command eq 'E' ) {
		my $it = e2char( $seq_arg );
		if ( defined $it ) {
			return $it;
		}
		else {
			carp "Unknown escape: E<$seq_arg>";
			return "E<$seq_arg>";
		}
	}

	# For all the other sequences, empty content produces no output.
	return if $seq_arg eq '';

	if ( $command eq 'B' or $command eq 'I' or $command eq 'S' ) {
		$seq_arg;
	}
	elsif ( $command eq 'C' or $command eq 'F' ) {

		# don't lose word-boundaries
		my $out = '';
		$out .= ' ' if s/^\s+//s;
		my $append;
		$append = 1 if s/\s+$//s;
		$out .= '_' if length $seq_arg;

		# which, if joined to another word, will set off the Perl-token alarm
		$out .= ' ' if $append;
		$out;
	}
	elsif ( $command eq 'L' ) {
		return $1 if m/^([^|]+)\|/s;
		'';
	}
	else {
		carp "Unknown sequence $command<$seq_arg>";
	}
}

#--------------------------------------------------------------------------

sub _treat_words {
	my ($self, $text) = @_;
	my $out = $self->stopwords->strip_stopwords( $text );
	if ( length $out ) {
		my $out_fh = $self->output_handle();
		print $out_fh wrap( '', '', $out ), "\n\n";
	}
	return;
}

#--------------------------------------------------------------------------

1;

# ABSTRACT: a formatter for spellchecking Pod

__END__

=pod

=encoding UTF-8

=head1 NAME

Pod::Spell - a formatter for spellchecking Pod

=head1 VERSION

version 1.15

=head1 SYNOPSIS

	use Pod::Spell;
	Pod::Spell->new->parse_from_file( 'File.pm' );

	Pod::Spell->new->parse_from_filehandle( $infile, $outfile );

Also look at L<podspell>

	% perl -MPod::Spell -e "Pod::Spell->new->parse_from_file(shift)" Thing.pm |spell |fmt

...or instead of piping to spell or C<ispell>, use C<E<gt>temp.txt>, and open
F<temp.txt> in your word processor for spell-checking.

=head1 DESCRIPTION

Pod::Spell is a Pod formatter whose output is good for
spellchecking.  Pod::Spell rather like L<Pod::Text|Pod::Text>, except that
it doesn't put much effort into actual formatting, and it suppresses things
that look like Perl symbols or Perl jargon (so that your spellchecking
program won't complain about mystery words like "C<$thing>"
or "C<Foo::Bar>" or "hashref").

This class provides no new public methods.  All methods of interest are
inherited from L<Pod::Parser|Pod::Parser> (which see).  The especially
interesting ones are C<parse_from_filehandle> (which without arguments
takes from STDIN and sends to STDOUT) and C<parse_from_file>.  But you
can probably just make do with the examples in the synopsis though.

This class works by filtering out words that look like Perl or any
form of computerese (like "C<$thing>" or "C<NE<gt>7>" or
"C<@{$foo}{'bar','baz'}>", anything in CE<lt>...E<gt> or FE<lt>...E<gt>
codes, anything in verbatim paragraphs (code blocks), and anything
in the stopword list.  The default stopword list for a document starts
out from the stopword list defined by L<Pod::Wordlist|Pod::Wordlist>,
and can be supplemented (on a per-document basis) by having
C<"=for stopwords"> / C<"=for :stopwords"> region(s) in a document.

=head1 METHODS

=head2 new

=head2 command

=head2 interior_sequence

=head2 textblock

=head2 verbatim

=head2 stopwords

	$self->stopwords->isa('Pod::WordList'); # true

=for :stopwords virtE<ugrave>

=head1 ENCODINGS

Pod::Parser, which Pod::Spell extends, is extremely naive about
character encodings.  The C<parse_from_file> method does not apply
any PerlIO encoding layer.  If your Pod file is encoded in UTF-8,
your data will be read incorrectly.

You should instead use C<parse_from_filehandle> and manage the input
and output layers yourself.

	binmode($_, ":utf8") for ($infile, $outfile);
	$my ps = Pod::Spell->new;
	$ps->parse_from_filehandle( $infile, $outfile );

If your output destination cannot handle UTF-8, you should set your
output handle to Latin-1 and tell Pod::Spell to strip out words
with wide characters.

	binmode($infile, ":utf8");
	binmode($outfile, ":encoding(latin1)");
	$my ps = Pod::Spell->new( no_wide_chars => 1 );
	$ps->parse_from_filehandle( $infile, $outfile );

=head1 ADDING STOPWORDS

You can add stopwords on a per-document basis with
C<"=for stopwords"> / C<"=for :stopwords"> regions, like so:

  =for stopwords  plok Pringe zorch   snik !qux
  foo bar baz quux quuux

This adds every word in that paragraph after "stopwords" to the
stopword list, effective for the rest of the document.  In such a
list, words are whitespace-separated.  (The amount of whitespace
doesn't matter, as long as there's no blank lines in the middle
of the paragraph.)  Plural forms are added automatically using
L<Lingua::EN::Inflect>. Words beginning with "!" are
I<deleted> from the stopword list -- so "!qux" deletes "qux" from the
stopword list, if it was in there in the first place.  Note that if
a stopword is all-lowercase, then it means that it's okay in I<any>
case; but if the word has any capital letters, then it means that
it's okay I<only> with I<that> case.  So a Wordlist entry of "perl"
would permit "perl", "Perl", and (less interestingly) "PERL", "pERL",
"PerL", et cetera.  However, a Wordlist entry of "Perl" catches
only "Perl", not "perl".  So if you wanted to make sure you said
only "Perl", never "perl", you could add this to the top of your
document:

  =for stopwords !perl Perl

Then all instances of the word "Perl" would be weeded out of the
Pod::Spell-formatted version of your document, but any instances of
the word "perl" would be left in (unless they were in a CE<lt>...> or
FE<lt>...> style).

You can have several "=for stopwords" regions in your document.  You
can even express them like so:

  =begin stopwords

  plok Pringe zorch

  snik !qux

  foo bar
  baz quux quuux

  =end stopwords

If you want to use EE<lt>...> sequences in a "stopwords" region, you
have to use ":stopwords", as here:

  =for :stopwords
  virtE<ugrave>

...meaning that you're adding a stopword of "virtE<ugrave>".  If
you left the ":" out, that would mean you were adding a stopword of
"virtEE<lt>ugrave>" (with a literal E, a literal <, etc), which
will have no effect, since  any occurrences of virtEE<lt>ugrave>
don't look like a normal human-language word anyway, and so would
be screened out before the stopword list is consulted anyway.

=head1 BUGS AND LIMITATIONS

=head2 finding stopwords defined with C<=for>

Pod::Spell makes a single pass over the POD.  Stopwords
must be added B<before> they show up in the POD.

=head2 finding the wordlist

Pod::Spell uses L<File::ShareDir::ProjectDistDir> if you're getting errors
about the wordlist being missing, chances are it's a problem with its
heuristics. Set C<PATH_ISDEV_DEBUG=1> or C<PATH_FINDDEV_DEBUG=1>, or both in
your environment for debugging, and then file a bug with
L<File::ShareDir::ProjectDistDir> if necessary.

=head1 HINT

If you feed output of Pod::Spell into your word processor and run a
spell-check, make sure you're I<not> also running a grammar-check -- because
Pod::Spell drops words that it thinks are Perl symbols, jargon, or
stopwords, this means you'll have ungrammatical sentences, what with
words being missing and all.  And you don't need a grammar checker
to tell you that.

=head1 SEE ALSO

L<Pod::Wordlist|Pod::Wordlist>

L<Pod::Parser|Pod::Parser>

L<podchecker|podchecker> also known as L<Pod::Checker|Pod::Checker>

L<perlpod|perlpod>, L<perlpodspec>

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/xenoterracide/pod-spell/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 CONTRIBUTORS

=over 4

=item *

David Golden <dagolden@cpan.org>

=item *

Kent Fredric <kentfredric@gmail.com>

=item *

Olivier Mengué <dolmen@cpan.org>

=back

=head1 AUTHORS

=over 4

=item *

Sean M. Burke <sburke@cpan.org>

=item *

Caleb Cushing <xenoterracide@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Caleb Cushing.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut
